SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,710 @@
+         !pai.isPullbackInResultOfAutodiffVJP,
+         pai.isPartialApplyOfReabstractionThunk,
+         pai.isSupportedClosure,
+         pai.callee.type.isNoEscapeFunction,
+         pai.callee.type.isThickFunction
+      {
+        conversionsAndReabstractions.pushIfNotVisited(pai)
+        possibleMarkDependenceBases.insert(pai)
+        haveUsedReabstraction = true
+        continue
+      }
+
+      // Look through `mark_dependence` on partial_apply[stack]
+      if let mdi = use.instruction as? MarkDependenceInst,
+         possibleMarkDependenceBases.contains(mdi.base),  
+         mdi.value == use.value,
+         mdi.value.type.isNoEscapeFunction,
+         mdi.value.type.isThickFunction
+      {
+        conversionsAndReabstractions.pushIfNotVisited(mdi)
+        continue
+      }
+
+      // Only add values used in `partial_apply`s as remaining uses
+      if use.instruction is PartialApplyInst {
+        remainingUses.pushIfNotVisited(currUse)
+        allDirectAndTransitiveUses.removeLast()
+      }
+    }
+  }
+
+  return haveUsedReabstraction
+}
+
+/// Finalizes the call sites for a given root closure by adding a corresponding `ClosureArgDescriptor`
+/// to all call sites where the closure is ultimately passed as an argument.
+private func finalizeCallSites(for rootClosure: SingleValueInstruction, callSiteMap: inout CallSiteMap, 
+                               allDirectAndTransitiveUses: [Instruction], 
+                               intermediateClosureArgDescs: [IntermediateClosureArgDescriptor], 
+                               _ context: FunctionPassContext) 
+{
+  var lifeRange = InstructionRange(begin: rootClosure, context)
+  defer {
+    lifeRange.deinitialize()
+  }
+  lifeRange.insert(contentsOf: allDirectAndTransitiveUses)
+
+  let closureInfo = ClosureInfo(closure: rootClosure, lifetimeFrontier: Array(lifeRange.ends))
+
+  intermediateClosureArgDescs
+    .reduce(into: [:]) { result, imClosureArgDesc in
+      result[imClosureArgDesc.callSite, default: []]
+        .append(imClosureArgDesc.intoClosureArgDescriptor(closureInfo: closureInfo))
+    }
+    .forEach { callSite, closureArgDescriptors in
+      if var callSiteDesc = callSiteMap[callSite] {
+        callSiteDesc.appendClosureArgDescriptors(closureArgDescriptors)
+        callSiteMap.update(key: callSite, value: callSiteDesc)
+      } else {
+        fatalError("Call site descriptor not found for call site: \(callSite)!")
+      }
+    }
+}
+
+private func isClosureApplied(in callee: Function, closureArgIndex index: Int) -> Bool {
+  func inner(_ callee: Function, _ index: Int, _ handledFuncs: inout Set<Function>) -> Bool {
+    let closureArg = callee.argument(at: index)
+
+    for use in closureArg.uses {
+      if let fai = use.instruction as? FullApplySite {
+        if fai.callee == closureArg {
+          return true
+        }
+
+        if let faiCallee = fai.referencedFunction,
+           faiCallee.isAvailableExternally,
```

Comment:
It's actually `!isAvailableExternally`. I corrected it in a local working version of this optimization upon testing, but forgot to propagate them here.

Reply:
okay, that makes more sense. Still, why can't we specialize external functions? The important thing is that the SIL is available

Reply:
> The important thing is that the SIL is available
Right, that is what matters. I guess I should directly check if the function's body exists instead of `isAvailableExternally`?

Reply:
I think so